### PR TITLE
Bump k8s.io/api, k8s.io/client-go, k8s.io/apimachinery from 0.22.7 to 0.23.6

### PR DESCRIPTION
**Pull request by mar4uk**
**What this PR does / why we need it**:
This PR bumps versions of k8s.io packages. This upgrade is needed to solve dependency issues when importing loki in other repositories where version of k8s.io is already upgraded. 
For example, when I try to use loki in https://github.com/open-telemetry/opentelemetry-collector-contrib I face dependencies inconsistencies in modules which uses newer version of k8s.io packages:
```
go: finding module for package k8s.io/api/batch/v2alpha1
go: finding module for package k8s.io/api/auditregistration/v1alpha1
go: finding module for package k8s.io/api/settings/v1alpha1
github.com/open-telemetry/opentelemetry-collector-contrib/internal/components imports
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/k8sclusterreceiver imports
	k8s.io/client-go/informers imports
	k8s.io/api/auditregistration/v1alpha1: module k8s.io/api@latest found (v0.23.6), but does not contain package k8s.io/api/auditregistration/v1alpha1
github.com/open-telemetry/opentelemetry-collector-contrib/internal/components imports
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/k8sclusterreceiver imports
	k8s.io/client-go/informers imports
	k8s.io/api/batch/v2alpha1: module k8s.io/api@latest found (v0.23.6), but does not contain package k8s.io/api/batch/v2alpha1
github.com/open-telemetry/opentelemetry-collector-contrib/internal/components imports
	github.com/open-telemetry/opentelemetry-collector-contrib/receiver/k8sclusterreceiver imports
	k8s.io/client-go/informers imports
	k8s.io/api/settings/v1alpha1: module k8s.io/api@latest found (v0.23.6), but does not contain package k8s.io/api/settings/v1alpha1
```
These packages: k8s.io/api/auditregistration/v1alpha1, k8s.io/api/batch/v2alpha1, k8s.io/api/settings/v1alpha1 are available in k8s.io@v.0.17.5, somehow loki brings this old version of k8s.io package when it uses k8s.io/api@v0.22.7

Updating k8s.io packages to v0.23.6 solves this dependency issue

**Special notes for your reviewer**:

**Checklist**
- [ ] Documentation added
- [ ] Tests updated
- [ ] Is this an important fix or new feature? Add an entry in the `CHANGELOG.md`.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/upgrading/_index.md`

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -312,3 +314,5 @@ replace github.com/bradfitz/gomemcache => github.com/themihai/gomemcache v0.0.0-
 replace github.com/thanos-io/thanos v0.22.0 => github.com/thanos-io/thanos v0.19.1-0.20211126105533-c5505f5eaa7d
 
 replace github.com/cloudflare/cloudflare-go => github.com/cyriltovena/cloudflare-go v0.27.1-0.20211118103540-ff77400bcb93
+
+exclude k8s.io/client-go v8.0.0+incompatible
```
Comment (owen-d) [inline go.mod:318]: I'm trying to understand this `exclude` directive -- can you help me understand what it helps with?

**Review by owen-d — COMMENTED** [review]
Hey, thanks for the PR! I'm not against bumping these versions if it's helpful. I restarted the build to see if it will pass, but left a question about the `exclude` directive.

File: go.mod (hunk 1 above)
Comment (mar4uk) [reply go.mod:318]: Adding exclude `k8s.io/client-go v8.0.0+incompatible` helps to keep version of github.com/influxdata/telegraf v1.16.3.
When I upgrade k8s.io/client-go to v0.23.6, go downgrades github.com/influxdata/telegraf to v0.10.1. When I upgrade telegraf back to v1.16.3, go upgrades k8s.io/client-go to v8.0.0+incompatible. But version v8.0.0+incompatible is actually older than v0.23.6.

`exclude k8s.io/client-go v8.0.0+incompatible` added to have both github.com/influxdata/telegraf v1.16.3 and k8s.io/client-go v0.23.6. 
I took this solution from prometheus https://github.com/prometheus/prometheus/blob/main/go.mod#L102-L103.

Not sure if we need to exclude other pre-go-mod kubernetes tags here. I decided to go with as small diff as possible in this PR

Comment (owen-d) [reply go.mod:318]: Instead of ignoring `k8s.io/client-go` to fix that, can we pin the telegraf version instead?

Comment (owen-d) [reply go.mod:318]: Ok, I think that's reasonable :+1:

**Review by owen-d — APPROVED** [review]

**Pull request merged**